File: v2LibrarySketch/Examples/Seq2SeqWithAttention.cpp
// A stacked LSTM network for classification, comprised of multiple stacked LSTM layers with self-stabilization
// This is representative of LSTM models used for Acoustic modelling in the ASR (automatic speech recognition) pipeline

#include "Trainer.h"
#include <assert.h>
#include <algorithm>

using namespace CNTK;

std::pair<FunctionPtr, FunctionPtr> LSTMPCellWithSelfStab(Variable input,
                                                          Variable prevOutput,
                                                          Variable prevCellState)
{
    assert(input.Shape().size() == 1);
    size_t inputDim = input.Shape()[0];

    size_t outputDim = prevOutput.Shape()[0];
    size_t cellDim = prevCellState.Shape()[0];

    auto Wxo = CNTK::Parameter(CNTK::RandomUniform({ cellDim, inputDim }, -0.5, 0.5), L"WxoParam");
    auto Wxi = CNTK::Parameter(CNTK::RandomUniform({ cellDim, inputDim }, -0.5, 0.5), L"WxiParam");
    auto Wxf = CNTK::Parameter(CNTK::RandomUniform({ cellDim, inputDim }, -0.5, 0.5), L"WxfParam");
    auto Wxc = CNTK::Parameter(CNTK::RandomUniform({ cellDim, inputDim }, -0.5, 0.5), L"WxcParam");

    auto Bo = CNTK::Parameter(CNTK::Constant({ cellDim }, 0.0), L"BoParam");
    auto Bc = CNTK::Parameter(CNTK::Constant({ cellDim }, 0.0), L"BcParam");
    auto Bi = CNTK::Parameter(CNTK::Constant({ cellDim }, 0.0), L"BiParam");
    auto Bf = CNTK::Parameter(CNTK::Constant({ cellDim }, 0.0), L"BfParam");

    auto Whi = CNTK::Parameter(CNTK::RandomUniform({ cellDim, outputDim }, -0.5, 0.5), L"WhiParam");
    auto Wci = CNTK::Parameter(CNTK::RandomUniform({ cellDim }, -0.5, 0.5), L"WciParam");

    auto Whf = CNTK::Parameter(CNTK::RandomUniform({ cellDim, outputDim }, -0.5, 0.5), L"WhfParam");
    auto Wcf = CNTK::Parameter(CNTK::RandomUniform({ cellDim }, -0.5, 0.5), L"WcfParam");

    auto Who = CNTK::Parameter(CNTK::RandomUniform({ cellDim, outputDim }, -0.5, 0.5), L"WhoParam");
    auto Wco = CNTK::Parameter(CNTK::RandomUniform({ cellDim }, -0.5, 0.5), L"WcoParam");

    auto Whc = CNTK::Parameter(CNTK::RandomUniform({ cellDim, outputDim }, -0.5, 0.5), L"WhcParam");

    auto Wmr = CNTK::Parameter(CNTK::RandomUniform({ outputDim, cellDim }, -0.5, 0.5), L"WmrParam");

    auto sWxo = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWxoParam");
    auto sWxi = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWxiParam");
    auto sWxf = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWxfParam");
    auto sWxc = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWxcParam");

    auto sWhi = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWhiParam");
    auto sWci = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWciParam");

    auto sWhf = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWhfParam");
    auto sWcf = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWcfParam");
    auto sWho = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWhoParam");
    auto sWco = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWcoParam");
    auto sWhc = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWhcParam");

    auto sWmr = CNTK::Parameter(CNTK::Constant({ 1, 1 }, 0.0), L"sWmrParam");

    auto expsWxo = CNTK::Exp(sWxo);
    auto expsWxi = CNTK::Exp(sWxi);
    auto expsWxf = CNTK::Exp(sWxf);
    auto expsWxc = CNTK::Exp(sWxc);

    auto expsWhi = CNTK::Exp(sWhi);
    auto expsWci = CNTK::Exp(sWci);

    auto expsWhf = CNTK::Exp(sWhf);
    auto expsWcf = CNTK::Exp(sWcf);
    auto expsWho = CNTK::Exp(sWho);
    auto expsWco = CNTK::Exp(sWco);
    auto expsWhc = CNTK::Exp(sWhc);

    auto expsWmr = CNTK::Exp(sWmr);

    auto Wxix = CNTK::Times(Wxi, CNTK::Scale(expsWxi, input));
    auto Whidh = CNTK::Times(Whi, CNTK::Scale(expsWhi, prevOutput));
    auto Wcidc = CNTK::DiagTimes(Wci, CNTK::Scale(expsWci, prevCellState));

    auto it = CNTK::Sigmoid(CNTK::Plus(CNTK::Plus(CNTK::Plus(Wxix, Bi), Whidh), Wcidc));

    auto Wxcx = CNTK::Times(Wxc, CNTK::Scale(expsWxc, input));
    auto Whcdh = CNTK::Times(Whc, CNTK::Scale(expsWhc, prevOutput));
    auto bit = CNTK::ElementTimes(it, CNTK::Tanh(CNTK::Plus(Wxcx, CNTK::Plus(Whcdh, Bc))));

    auto Wxfx = CNTK::Times(Wxf, CNTK::Scale(expsWxf, input));
    auto Whfdh = CNTK::Times(Whf, CNTK::Scale(expsWhf, prevOutput));
    auto Wcfdc = CNTK::DiagTimes(Wcf, CNTK::Scale(expsWcf, prevCellState));

    auto ft = CNTK::Sigmoid(CNTK::Plus(CNTK::Plus(CNTK::Plus(Wxfx, Bf), Whfdh), Wcfdc));

    auto bft = CNTK::ElementTimes(ft, prevCellState);

    auto ct = CNTK::Plus(bft, bit);

    auto Wxox = CNTK::Times(Wxo, CNTK::Scale(expsWxo, input));
    auto Whodh = CNTK::Times(Who, CNTK::Scale(expsWho, prevOutput));
    auto Wcoct = CNTK::DiagTimes(Wco, CNTK::Scale(expsWco, ct));

    auto ot = CNTK::Sigmoid(CNTK::Plus(CNTK::Plus(CNTK::Plus(Wxox, Bo), Whodh), Wcoct));

    auto mt = CNTK::ElementTimes(ot, CNTK::Tanh(ct));

    return { CNTK::Times(Wmr, CNTK::Scale(expsWmr, mt)), ct };
}

FunctionPtr EncoderSubNet(size_t inputDim, size_t cellDim, size_t hiddenDim)
{
    Variable sourceInput({ inputDim }, L"Source");

    auto outputPlaceholder = Variable({ hiddenDim }, L"outputPlaceHolder");
    auto dh = CNTK::PastValue(CNTK::Constant(outputPlaceholder.Shape(), 0.0), outputPlaceholder, L"OutputPastValue");
    auto ctPlaceholder = Variable({ cellDim }, L"ctPlaceHolder");
    auto dc = CNTK::PastValue(CNTK::Constant(ctPlaceholder.Shape(), 0.0), ctPlaceholder, L"CellPastValue");

    auto LSTMCell = LSTMPCellWithSelfStab(sourceInput, dh, dc);

    // Form the recurrence loop by connecting the output and cellstate back to the inputs of the respective PastValue nodes
    return CNTK::Composite(LSTMCell.first, { { outputPlaceholder, LSTMCell.first }, { ctPlaceholder, LSTMCell.second } });
}

// Attention as described in http://arxiv.org/pdf/1412.7449v3.pdf
FunctionPtr Attention(Variable encoderStates, Variable decoderState)
{
    size_t encoderStateDim = encoderStates.Shape()[0];
    size_t decoderStateDim = decoderState.Shape()[0];
    size_t attentionDim = decoderStateDim;
    auto encoderStateProjParams = CNTK::Parameter(CNTK::RandomUniform({ attentionDim, encoderStateDim }, -0.5, 0.5), L"AttentionEncoderParams");
    auto encoderProj = CNTK::Times(encoderStateProjParams, encoderStates);
    auto decoderStateProjParams = CNTK::Parameter(CNTK::RandomUniform({ attentionDim, decoderStateDim }, -0.5, 0.5), L"AttentionDecoderParams");
    auto decoderProj = CNTK::Times(decoderStateProjParams, decoderState);

    auto reductionParams = CNTK::Parameter(CNTK::RandomUniform({ 1, attentionDim }, -0.5, 0.5), L"AttentionReductionParams");
    // The Plus operation below broadcasts along the column dimension of the projected encoder state
    auto u = CNTK::Times(reductionParams, CNTK::Tanh(CNTK::Plus(decoderProj, encoderProj)));

    // Perform a Softmax along the column dimension of 'u' to obtain the weight vector for the attention
    auto attentionWeights = CNTK::Softmax(u, 1 /*axis*/);

    return CNTK::Times(encoderStates, attentionWeights);
}

// Note that this is just for training where we use an input target sentence to drive the decoder instead of its own output from previous step
FunctionPtr DecoderWithAttention(Variable encoderStates, size_t outputDim, size_t cellDim)
{
    // The decoder is a recurrent network that uses attention over the hidden states of the encoder emitted for each step of the source sequence

    Variable targetInput({ outputDim }, L"Target");

    // Compute the context vector using attention over the encoderStates
    auto ctPlaceholder = Variable({ cellDim }, L"ctPlaceHolder");
    auto dc = CNTK::PastValue(CNTK::Constant(ctPlaceholder.Shape(), 0.0), ctPlaceholder, L"CellPastValue");

    auto context = Attention(encoderStates, dc);
    auto LSTMCell = LSTMPCellWithSelfStab(context, targetInput, dc);

    // Form the recurrence loop by connecting the cellstate back to the input of the PastValue node
    return CNTK::Composite(LSTMCell.first, { { ctPlaceholder, LSTMCell.second } });
}

FunctionPtr EncoderDecoderWithAttention(size_t inputDim, size_t cellDim, size_t hiddenDim, size_t outputDim)
{
    auto encoderFunction = EncoderSubNet(inputDim, cellDim, hiddenDim);

    // Each Variable has a Shape (denoting the shape of a sample) and one implicit sequence dimension(s) denoting the fixed or variable lengths of
    // the sequence that the Variable denotes.
    // We reshape the encoder state sequence to fold the sequence axis into the sample shape so that it turns from being a sequence of states to
    // a single sample consisting of variable number of columns corresponding to the length of the source sequence. We then perform batch algebraic and reduction
    // operations over the encoder state inside the decoder recurrence loop.
    NDShape newShape = { 1, INFERRED_DIMENSION };
    newShape.insert(newShape.end(), encoderFunction->Output().Shape().begin(), encoderFunction->Output().Shape().end());
    auto encoderStates = CNTK::Reshape(encoderFunction, -1, (int)encoderFunction->Output().Shape().size(), newShape);

    return DecoderWithAttention(encoderStates, outputDim, cellDim);
}

void TrainEncoderDecoder(ReaderPtr trainingDataReader)
{
    StreamDescription sourceSequenceStreamDesc = CNTK::GetStreamDescription(trainingDataReader, L"Source");
    const size_t inputDim = sourceSequenceStreamDesc.m_sampleLayout[0];

    StreamDescription targetSequenceStreamDesc = CNTK::GetStreamDescription(trainingDataReader, L"Target");
    const size_t outputDim = targetSequenceStreamDesc.m_sampleLayout[0];

    const size_t cellDim = 1024;
    const size_t hiddenDim = 512;
    auto encoderDecoderNetOutputFunction = EncoderDecoderWithAttention(inputDim, cellDim, hiddenDim, outputDim);

    Variable labelsVar = Variable({ outputDim }, L"Target");
    auto trainingLossFunction = CNTK::CrossEntropyWithSoftmax(encoderDecoderNetOutputFunction, labelsVar, L"lossFunction");
    auto trainingNet = trainingLossFunction;

    size_t momentumTimeConstant = 1024;
    double learningRatePerSample = 0.05;

    // Train for 100000 samples; checkpoint every 10000 samples
    TrainingControlPtr driver = CNTK::BasicTrainingControl(100000, 10000, { L"EncoderDecoderWithAttn.net", L"EncoderDecoderWithAttn.ckp" });
    Trainer encoderDecoderTrainer(trainingNet, trainingLossFunction, { CNTK::SGDLearner(trainingNet->Parameters(), learningRatePerSample, momentumTimeConstant) });

    std::unordered_map<Variable, StreamDescription> modelArgumentToReaderStreamMap = { { encoderDecoderNetOutputFunction->Argument(), sourceSequenceStreamDesc }, { labelsVar, targetSequenceStreamDesc } };
    encoderDecoderTrainer.Train(trainingDataReader, modelArgumentToReaderStreamMap, driver);
}

File: v2LibrarySketch/Python/examples/seq2seq_with_attention.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
#

# A stacked LSTM network for classification, comprised of multiple stacked LSTM layers with self-stabilization
# This is representative of LSTM models used for Acoustic modelling in the ASR (automatic speech recognition) pipeline

from ..cntk import *

def LSTMP_cell_with_self_stabilization(input: Variable, prevOutput: Variable, prevCellState: Variable) -> Tuple[Function, Function]:
    assert(len(input.shape()) == 1)
    inputDim = input.shape()[0]

    outputDim = prevOutput.shape()[0]
    cellDim = prevCellState.shape()[0]

    Wxo = parameter(random_uniform([ cellDim, inputDim ], -0.5, 0.5), "WxoParam")
    Wxi = parameter(random_uniform([ cellDim, inputDim ], -0.5, 0.5), "WxiParam")
    Wxf = parameter(random_uniform([ cellDim, inputDim ], -0.5, 0.5), "WxfParam")
    Wxc = parameter(random_uniform([ cellDim, inputDim ], -0.5, 0.5), "WxcParam")

    Bo = parameter(constant([ cellDim ], 0.0), "BoParam")
    Bc = parameter(constant([ cellDim ], 0.0), "BcParam")
    Bi = parameter(constant([ cellDim ], 0.0), "BiParam")
    Bf = parameter(constant([ cellDim ], 0.0), "BfParam")

    Whi = parameter(random_uniform([ cellDim, outputDim ], -0.5, 0.5), "WhiParam")
    Wci = parameter(random_uniform([ cellDim ], -0.5, 0.5), "WciParam")

    Whf = parameter(random_uniform([ cellDim, outputDim ], -0.5, 0.5), "WhfParam")
    Wcf = parameter(random_uniform([ cellDim ], -0.5, 0.5), "WcfParam")

    Who = parameter(random_uniform([ cellDim, outputDim ], -0.5, 0.5), "WhoParam")
    Wco = parameter(random_uniform([ cellDim ], -0.5, 0.5), "WcoParam")

    Whc = parameter(random_uniform([ cellDim, outputDim ], -0.5, 0.5), "WhcParam")

    Wmr = parameter(random_uniform([ outputDim, cellDim ], -0.5, 0.5), "WmrParam")

    sWxo = parameter(0.0, "sWxoParam")
    sWxi = parameter(0.0, "sWxiParam")
    sWxf = parameter(0.0, "sWxfParam")
    sWxc = parameter(0.0, "sWxcParam")

    sWhi = parameter(0.0, "sWhiParam")
    sWci = parameter(0.0, "sWciParam")

    sWhf = parameter(0.0, "sWhfParam")
    sWcf = parameter(0.0, "sWcfParam")
    sWho = parameter(0.0, "sWhoParam")
    sWco = parameter(0.0, "sWcoParam")
    sWhc = parameter(0.0, "sWhcParam")

    sWmr = parameter(0.0, "sWmrParam")

    expsWxo = exp(sWxo)
    expsWxi = exp(sWxi)
    expsWxf = exp(sWxf)
    expsWxc = exp(sWxc)

    expsWhi = exp(sWhi)
    expsWci = exp(sWci)

    expsWhf = exp(sWhf)
    expsWcf = exp(sWcf)
    expsWho = exp(sWho)
    expsWco = exp(sWco)
    expsWhc = exp(sWhc)

    expsWmr = exp(sWmr)

    Wxix = times(Wxi, expsWxi * input)
    Whidh = times(Whi, expsWhi * prevOutput)
    Wcidc = Wci * (expsWci * prevCellState)

    it = sigmoid((((Wxix + Bi) + Whidh) + Wcidc))

    Wxcx = times(Wxc, expsWxc * input)
    Whcdh = times(Whc, expsWhc * prevOutput)
    bit = it * tanh(Wxcx + (Whcdh + Bc))

    Wxfx = times(Wxf, expsWxf * input)
    Whfdh = times(Whf, expsWhf * prevOutput)
    Wcfdc = Wcf * (expsWcf * prevCellState)

    ft = sigmoid((((Wxfx + Bf) + Whfdh) + Wcfdc))

    bft = ft * prevCellState

    ct = bft + bit

    Wxox = times(Wxo, expsWxo * input)
    Whodh = times(Who, expsWho * prevOutput)
    Wcoct = Wco * (expsWco * ct)

    ot = sigmoid((((Wxox + Bo) + Whodh) + Wcoct))

    mt = ot * tanh(ct)

    return ( times(Wmr, (expsWmr * mt)), ct )

def encoder_subnet(sourceInput: Variable, cellDim: int, hiddenDim: int) -> Function:
    outputPlaceholder = Variable([ hiddenDim ], "outputPlaceHolder")
    dh = past_value(0.0, outputPlaceholder, "OutputPastValue")
    ctPlaceholder = Variable([ cellDim ], "ctPlaceHolder")
    dc = past_value(0.0, ctPlaceholder, "CellPastValue")

    LSTMCell = LSTMP_cell_with_self_stabilization(sourceInput, dh, dc)

    # Form the recurrence loop by connecting the output and cellstate back to the inputs of the respective past_value nodes
    return composite(LSTMCell[0], {outputPlaceholder : LSTMCell[0], ctPlaceholder : LSTMCell[1]})

# Attention as described in http://arxiv.org/pdf/1412.7449v3.pdf
def attention(encoderState: Variable, decoderState: Variable) -> Function:
    attentionDim = decoderState.shape()
    encoderStateProjParams = parameter(random_uniform(attentionDim.append_shape(encoderState.shape()), -0.5, 0.5), "AttentionEncoderParams")
    encoderProj = times(encoderStateProjParams, encoderState)
    decoderStateProjParams = parameter(random_uniform(attentionDim.append_shape(decoderState.shape()), -0.5, 0.5), "AttentionDecoderParams")
    decoderProj = times(decoderStateProjParams, decoderState)

    reductionParams = parameter(random_uniform(NDShape([ 1 ]).append_shape(attentionDim), -0.5, 0.5), "AttentionReductionParams")

    # The Plus operation below broadcasts along the column dimension of the projected encoder state
    u = times(reductionParams, tanh(decoderProj + encoderProj))

    # Perform a Softmax along the sequence axis of encoderState to obtain the weight vector for the attention
    encoderStateDynamicAxis = encoderState.dynamic_axes()[0]
    attentionWeights = softmax(u, encoderStateDynamicAxis)

    # Now we multiply the encoderState with the attention state and then sum along the sequence axis of encoderState
    return sum(attentionWeights * encoderState, encoderStateDynamicAxis)

# Note that this is just for training where we use an input target sentence to drive the decoder instead of its own output from previous step
def decoder_with_attention(targetInput: Variable, encoderStates: Variable, cellDim: int) -> Function:
    # The decoder is a recurrent network that uses attention over the hidden states of the encoder for each step of the source sequence

    # Compute the context vector using attention over the encoderStates
    outputPlaceholder = Variable([ INFERRED_DIMENSION ], "outputPlaceholder")
    dh = past_value(0.0, outputPlaceholder, "OutputPastValue")

    context = attention(encoderStates, dh)
    LSTMCell = LSTMP_cell_with_self_stabilization(context, targetInput, dh)

    # Form the recurrence loop by connecting the cellstate back to the input of the past_value node
    return composite(LSTMCell[0], {outputPlaceholder : LSTMCell[0]})

def encoder_decoder_with_attention(sourceInput: Variable, targetInput: Variable, cellDim: int, hiddenDim: int) -> Function:
    encoderFunction = encoder_subnet(sourceInput, cellDim, hiddenDim)
    return decoder_with_attention(targetInput, encoderFunction, cellDim)

def train_encoder_decoder(trainingDataMinibatchSource: MinibatchSource):
    sourceSequenceStreamDesc = get_stream_description(trainingDataMinibatchSource, "Source")
    inputDim = sourceSequenceStreamDesc.m_sampleLayout[0]
    sourceInput = Variable([ inputDim ], AxisId.new_dynamic_axis("Source"), "Source")

    targetSequenceStreamDesc = get_stream_description(trainingDataMinibatchSource, "Target")
    outputDim = targetSequenceStreamDesc.m_sampleLayout[0]
    targetInput = Variable([ outputDim ], AxisId.new_dynamic_axis("Target"), "Target")

    cellDim = 1024
    hiddenDim = 512
    encoderDecoderNetOutputFunction = encoder_decoder_with_attention(sourceInput, targetInput, cellDim, hiddenDim)

    trainingLossFunction = cross_entropy_with_softmax(encoderDecoderNetOutputFunction, targetInput, "lossFunction")
    trainingNet = trainingLossFunction

    momentumTimeConstant = 1024
    learningRatePerSample = 0.05

    # Train for 100000 samples checkpoint every 10000 samples
    driver = basic_training_control(100000, 10000, [ "EncoderDecoderWithAttn.net", "EncoderDecoderWithAttn.ckp" ])
    encoderDecoderTrainer = Trainer(trainingNet, trainingLossFunction, {sgd_learner(trainingNet.parameters(), learningRatePerSample, momentumTimeConstant)})

    modelArgumentToMinibatchSourceStreamMap = {sourceInput : sourceSequenceStreamDesc, targetInput : targetSequenceStreamDesc}
    encoderDecoderTrainer.train(trainingDataMinibatchSource, modelArgumentToMinibatchSourceStreamMap, driver)
